src/app/InteractionModelEngine.cpp

```diff
@@ -530,5 +530,125 @@ bool InteractionModelEngine::IsOverlappedAttributePath(ClusterInfo & aAttributeP
     return false;
 }
 
+void InteractionModelEngine::DispatchCommand(CommandHandler & apCommandObj, const ConcreteCommandPath & aCommandPath,
+                                             TLV::TLVReader & apPayload)
+{
+    CommandHandlerInterface * handler = FindCommandHandler(aCommandPath.mEndpointId, aCommandPath.mClusterId);
+
+    if (handler)
+    {
+        CommandHandlerInterface::HandlerContext context(apCommandObj, aCommandPath, apPayload);
+        handler->InvokeCommand(context);
+
+        //
+        // If the command was handled, don't proceed any further and return successfully.
+        //
+        if (context.mCommandHandled)
+        {
+            return;
+        }
+    }
+
+    DispatchSingleClusterCommand(aCommandPath, apPayload, &apCommandObj);
+}
+
+bool InteractionModelEngine::CommandExists(const ConcreteCommandPath & aCommandPath)
+{
+    CommandHandlerInterface * handler = FindCommandHandler(aCommandPath.mEndpointId, aCommandPath.mClusterId);
+    if (handler)
+    {
+        return true;
```

Comment:
But... the actual check is supposed to be for whether the command exists, not whether the cluster exists.  Either the check needs to be changed in terms of its semantics (and then a command not existing would get determined via trying to dispatch to the (existing) cluster and failing) or something.

Reply:
Well, it depends what behavior we're after.

It appears to me that the PR as written catches all commands for a cluster instance, or none.  It sounds like the suggestion is that it would be better to be able to register by each `(optional<endpointId>, clusterId, commandId)` individually.

Reply:
Ah, I see.  This is claiming that a command _might_ exist.  You will only know for sure by actually executing the handler and checking mCommandHandled.

Is there a reason this can't work?  What will happen if the candidate handler is executed and it turns out it doesn't handle the command?  I think in that case DispatchSingleClusterCommand will be called as fallback.  Do things blow up or fail sanely / gracefully if this doesn't have a handler?

Reply:
Thinking on this some more, it seems like this feature is just incompatible with CommandExists in the first place because a the dynamic command handler is cluster-wide and might handle a command or not.

Is there a reason we need CommandExists at all?  Why can't we dispatch to the handlers, and if somebody handles it, we're done?

Reply:
Dispatching via DispatchSingleClusterCommand performs the following checks via just code flow (switch statements):

1. The cluster identified by the cluster id in the path is generally enabled in the ZAP config.
2. The command identified by the command id in the path is generally enabled in the ZAP config.

What it does NOT check is that (a) the cluster is enabled on the specific endpoint in the path and (b) that the command is enabled on the specific cluster instance on that endpoint.

The existing `ServerClusterCommandExists` code checks (a) but not (b)....  For correctness we should be checking both (a) and (b) somewhere: otherwise we can end up handling a command targeted at an endpoint where that command is not enabled, which is presumably not something the app logic expects.

Reply:
Well then it seems like we want (a) and (b) checks inside `DispatchSingleClusterCommand`, as well as a guard for `mCommandHandled` already true so we can just call it blindly.

Is there a reason that `ServerClusterCommandExists` and `DispatchSingleClusterCommand` need to be separate functions?

Reply:
There is not per se, no.  We could have `DispatchSingleClusterCommand` do the checking itself, and probably should....  Note that right now we simply don't store the data to do check (b) anywhere at all, but we need to fix that anyway to implement the global cluster attributes...

Reply:
I don't think b) should happen at command dispatch time, in either `DispatchSingleClusterCommand` or `InvokeCommand`. Like we do with attributes, we should be building up command metadata descriptors for each endpoint and cluster in the device, that are then parsed fairly early on in the command parsing chain **before** we get around to actually dispatching the command. The means by which these descriptors are 'registered' with the stack (either statically, or dynamically) needs to be figured out separately. This will then 

In this PR however, I focused on maintaining parity with the existing implementation. Yes, `CommandExists`'s expectation of checking whether a command exists isn't what _actually_ any of this logic is doing - both the existing logic and `FindCommandHandler` here are only validating premise a). So, in that regard, I think they are both equally broken, but, suffices for now.

Once we get b) implemented, we should take out this logic.

Reply:
> In this PR however, I focused on maintaining parity with the existing implementation.

But the check here does not have parity, in the sense that it does not implement (a) correctly...

Reply:
Thinking about this a bit more, arguably, even in the case where someone has registered a `CommandHandlerInterface`, that presumably, they would be expected to register their cluster + endpoint using `emberAfSetDynamicEndpoint`. So, just calling `ServerClusterCommandExists` _should_ be sufficient to cover both cases right?

Reply:
Yes, that makes sense to me.